### PR TITLE
Feature/maintenance page

### PR DESCRIPTION
**Pull request by switchflip**
### Whats happening in here.
- `deploy.rb` will now setup a new maintenance page if the attribute `enable_maintance_page` is set to `true`.

### CODE REVIEW

**Comment by D1plo1d** [comment]
To summarize my feedback:

* remove the nginx.conf deletion because `template` will overwrite the file any ways
* remove the files/folders deletion in else. These are fine to keep around/take up negligible space/just add to chef code complexity as long as we're changing the nginx.conf.
* merge maintenance.conf.erb and nginx.conf.erb and use an if statement to toggle maintenance mode
* move test code to `before :all` and `after :all` blocks inside `describe` blocks
* change the `maintenance page is disabled` test to skip the setup and instead just run a deploy to emulate what we would run on production.

**Comment by switchflip** [comment]
Hey Rob, I've finished rewriting the setup and deploy recipes recipes as well as the tests (tests passing).

File: recipes/deploy.rb [hunk 1]
```diff
@@ -1,17 +1,62 @@
...
+  template '/var/www/cloud.touchbistro.com/maintenance/index.html' do
+    action  :create
+    source  'maintenance_page.html'
+    owner   'root'
+    group   'root'
+    mode    '0744'
+  end
+else
+  # regular operations
+  template '/etc/nginx/nginx.conf' do
```
Comment (D1plo1d) [inline recipes/deploy.rb]: Could this and line 20 be moved out of the if/else and merged?

Comment (switchflip) [inline recipes/deploy.rb]: I'm not sure, I know exactly what you mean?

Comment (D1plo1d) [inline recipes/deploy.rb]: so both like 40 and line 20 are creating `/etc/nginx/nginx.conf`. They have a few different variables passed in (it actually seems like a few are missing from line 40 that should be there for certs to work) and that's it. They look like duplicate code on both sides of an if/else that could be de-duplicated by moving it outside of the if/else statement.

File: test/integration/setup/serverspec/enable_maintenance_spec.rb [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    end
+
+    it "should return content from touchbistro's maintenance page" do
+      resp = get
+      expect(resp.body).to include('TouchBistro')
+    end
+  end
+
+  describe "maintenance page is disabled" do
+    before :each do
```
Comment (D1plo1d) [inline test/integration/setup/serverspec/enable_maintenance_spec.rb]: you could save a chef run by changing this to `:all`. It should have the same result.

File: recipes/deploy.rb (hunk 1 above)
Comment (switchflip) [inline recipes/deploy.rb]: Okay... working on that now.

File: test/integration/setup/serverspec/enable_maintenance_spec.rb (hunk 2 above)
Comment (switchflip) [inline test/integration/setup/serverspec/enable_maintenance_spec.rb]: Nice.

Comment (D1plo1d) [inline test/integration/setup/serverspec/enable_maintenance_spec.rb]: also we're still missing the line that would make this a standalone test that does not presume an earlier state from pervious tests! Please change this to

```ruby
[true, false].each do |maintenance_enabled|
     maintenance_page maintenance_enabled
     run_deploy
end
```

I realize this looks like duplication of behviour that you should logically expect from your other tests but the reality is that tests change regularly and tests that depend on side effects from other tests are a very big problem down the road. You end up in scenarios where you change test A and test Z starts breaking for no obvious reason.

File: recipes/deploy.rb (hunk 1 above)
Comment (D1plo1d) [inline recipes/deploy.rb]: thanks!

File: test/integration/setup/serverspec/enable_maintenance_spec.rb (hunk 2 above)
Comment (switchflip) [inline test/integration/setup/serverspec/enable_maintenance_spec.rb]: That makes a lot of sense. Cool, going to change now.

**Comment by switchflip** [comment]
Eh Rob, finished making all those changes and the tests are passing. Thanks again!

**Comment by D1plo1d** [comment]
Looks great! Thanks again for all your work reworking this PR!

**Pull request merged**